File: python/context_engineering/quality.py
"""Quality module: heuristic context quality metrics.

Computes density, diversity, freshness, and redundancy without LLM calls.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import List, Set

from .core import ContextItem, ContextPack, estimate_tokens


@dataclass
class ContextQuality:
    """Quality metrics for a context pack or item set."""

    item_count: int
    total_tokens: int
    density: float
    diversity: float
    freshness: float
    redundancy: float
    overall: float


def analyze_context(items: List[ContextItem]) -> ContextQuality:
    """Analyze the quality of a set of context items.

    Computes density, diversity, freshness, and redundancy metrics
    without requiring an LLM call.

    Args:
        items: Context items to analyze.

    Returns:
        ContextQuality with all metrics computed.
    """
    if not items:
        return ContextQuality(
            item_count=0,
            total_tokens=0,
            density=0.0,
            diversity=0.0,
            freshness=0.0,
            redundancy=0.0,
            overall=0.0,
        )

    total_tokens = sum(item.tokens or estimate_tokens(item.content) for item in items)

    # Density: unique words / total tokens
    all_words: List[str] = []
    for item in items:
        words = [w for w in item.content.lower().split() if w]
        all_words.extend(words)
    unique_words: Set[str] = set(all_words)
    density = min(len(unique_words) / max(total_tokens, 1), 1.0)

    # Diversity: unique bigrams / total bigrams
    bigrams: Set[str] = set()
    total_bigram_count = 0
    for item in items:
        words = [w for w in item.content.lower().split() if w]
        for i in range(len(words) - 1):
            bigrams.add(f"{words[i]} {words[i + 1]}")
            total_bigram_count += 1
    diversity = min(len(bigrams) / total_bigram_count, 1.0) if total_bigram_count > 0 else 0.0

    # Freshness: fraction of items with recency > 5 (on 0-10 scale)
    fresh_count = sum(1 for item in items if (item.recency or 0) > 5)
    freshness = fresh_count / len(items)

    # Redundancy: pairwise Jaccard similarity
    item_word_sets = [
        set(w for w in item.content.lower().split() if len(w) > 2) for item in items
    ]
    total_overlap = 0.0
    pair_count = 0
    for i in range(len(item_word_sets)):
        for j in range(i + 1, len(item_word_sets)):
            a = item_word_sets[i]
            b = item_word_sets[j]
            intersection = len(a & b)
            union = len(a | b)
            if union > 0:
                total_overlap += intersection / union
                pair_count += 1
    redundancy = total_overlap / pair_count if pair_count > 0 else 0.0

    # Overall: weighted combination
    overall = round(
        density * 0.25 + diversity * 0.25 + freshness * 0.20 + (1 - redundancy) * 0.30,
        2,
    )

    return ContextQuality(
        item_count=len(items),
        total_tokens=total_tokens,
        density=round(density, 3),
        diversity=round(diversity, 3),
        freshness=round(freshness, 3),
        redundancy=round(redundancy, 3),
        overall=overall,
    )


def analyze_context_pack(pack: ContextPack) -> ContextQuality:
    """Analyze a ContextPack directly."""
    return analyze_context(pack.selected)

File: packages/ce-core/src/quality.ts
import type { ContextItem, ContextPack } from "./types.js";
import { estimateTokens } from "./estimate.js";

/**
 * Quality metrics for a context pack or item set.
 */
export interface ContextQuality {
  /** Items analyzed */
  itemCount: number;
  /** Total tokens */
  totalTokens: number;
  /** Information density: unique words per token (0-1, higher = more dense) */
  density: number;
  /** Topic diversity: ratio of unique bigrams to total bigrams (0-1) */
  diversity: number;
  /** Freshness: fraction of items with recency > 5 on a 0-10 scale (0-1) */
  freshness: number;
  /** Redundancy: estimated content overlap between items (0-1, lower = better) */
  redundancy: number;
  /** Overall quality score (weighted combination, 0-1) */
  overall: number;
}

/**
 * Compute Jaccard similarity between two word sets.
 */
function jaccardSimilarity(a: Set<string>, b: Set<string>): number {
  let intersection = 0;
  for (const word of a) {
    if (b.has(word)) intersection++;
  }
  const union = a.size + b.size - intersection;
  return union > 0 ? intersection / union : 0;
}

/**
 * Analyze the quality of a set of context items.
 *
 * Computes density, diversity, freshness, and redundancy metrics
 * without requiring an LLM call — uses heuristic analysis.
 *
 * @param items - Context items to analyze
 * @returns Quality metrics including density, diversity, freshness, and redundancy
 *
 * @example
 * ```ts
 * const packed = await pack(items, budget);
 * const quality = analyzeContext(packed.selected);
 * console.log(quality.overall); // 0.82
 * if (quality.redundancy > 0.3) {
 *   console.warn("High redundancy — consider deduplication");
 * }
 * ```
 */
export function analyzeContext(items: ContextItem[]): ContextQuality {
  if (items.length === 0) {
    return {
      itemCount: 0,
      totalTokens: 0,
      density: 0,
      diversity: 0,
      freshness: 0,
      redundancy: 0,
      overall: 0,
    };
  }

  // Pre-compute word arrays once per item to avoid repeated splitting (M14 fix)
  const itemWords = items.map(item =>
    item.content
      .toLowerCase()
      .split(/\s+/)
      .filter(w => w.length > 0)
  );

  const totalTokens = items.reduce(
    (sum, item) => sum + (item.tokens ?? estimateTokens(item.content)),
    0
  );

  // Density: unique words / total tokens
  const uniqueWords = new Set<string>();
  for (const words of itemWords) {
    for (const w of words) {
      uniqueWords.add(w);
    }
  }
  const density = Math.min(uniqueWords.size / Math.max(totalTokens, 1), 1);

  // Diversity: unique bigrams / total bigrams
  const bigrams = new Set<string>();
  let totalBigramCount = 0;
  for (const words of itemWords) {
    for (let i = 0; i < words.length - 1; i++) {
      bigrams.add(`${words[i]} ${words[i + 1]}`);
      totalBigramCount++;
    }
  }
  const diversity =
    totalBigramCount > 0 ? Math.min(bigrams.size / totalBigramCount, 1) : 0;

  // Freshness: fraction of items with recency above midpoint (> 5 on 0-10 scale)
  const freshCount = items.filter(item => (item.recency ?? 0) > 5).length;
  const freshness = freshCount / items.length;

  // Redundancy: pairwise word overlap using Jaccard similarity.
  // For large item sets (> 100), sample pairs to avoid O(n^2) blowup.
  const itemWordSets = itemWords.map(
    words => new Set(words.filter(w => w.length > 2))
  );

  let totalOverlap = 0;
  let pairCount = 0;
  const n = itemWordSets.length;
  const MAX_PAIRS = 5000;
  const totalPossiblePairs = (n * (n - 1)) / 2;

  if (totalPossiblePairs <= MAX_PAIRS) {
    // Exhaustive comparison for small sets
    for (let i = 0; i < n; i++) {
      for (let j = i + 1; j < n; j++) {
        totalOverlap += jaccardSimilarity(itemWordSets[i], itemWordSets[j]);
        pairCount++;
      }
    }
  } else {
    // Sample random pairs for large sets to stay O(n)
    // Use deterministic sampling based on item count for reproducibility
    const sampleCount = Math.min(MAX_PAIRS, n * 10);
    for (let s = 0; s < sampleCount; s++) {
      // Simple deterministic pair selection using modular arithmetic
      const i = s % n;
      const j = (i + 1 + Math.floor(s / n)) % n;
      if (i !== j) {
        totalOverlap += jaccardSimilarity(itemWordSets[i], itemWordSets[j]);
        pairCount++;
      }
    }
  }
  const redundancy = pairCount > 0 ? totalOverlap / pairCount : 0;

  // Overall: weighted combination
  const overall =
    Math.round(
      (density * 0.25 +
        diversity * 0.25 +
        freshness * 0.2 +
        (1 - redundancy) * 0.3) *
        100
    ) / 100;

  return {
    itemCount: items.length,
    totalTokens,
    density: Math.round(density * 1000) / 1000,
    diversity: Math.round(diversity * 1000) / 1000,
    freshness: Math.round(freshness * 1000) / 1000,
    redundancy: Math.round(redundancy * 1000) / 1000,
    overall,
  };
}

/**
 * Analyze a ContextPack directly.
 *
 * @param pack - The context pack to analyze
 * @returns Quality metrics for the pack's selected items
 *
 * @example
 * ```ts
 * const packed = await pack(items, { maxTokens: 4000 });
 * const quality = analyzeContextPack(packed);
 * console.log(quality.overall);
 * ```
 */
export function analyzeContextPack(pack: ContextPack): ContextQuality {
  return analyzeContext(pack.selected);
}
